src/mon/AuthMonitor.cc

```diff
@@ -1632,32 +1633,37 @@ bool AuthMonitor::prepare_command(MonOpRequestRef op)
 
       mds_cap_string += mds_cap_string.empty() ? "" : ", ";
       mds_cap_string += "allow " + cap;
-      if (path != "/") {
-	mds_cap_string += " path=" + path;
+
+      if (filesystem != "*" && filesystem != "all") {
+	auto fs = mon->mdsmon()->get_fsmap().get_filesystem(filesystem);
+	if (!fs) {
+	  ss << "filesystem " << filesystem << " does not exist.";
+	  err = -EINVAL;
+	  goto done;
+	} else {
+	  mds_cap_string += " fsname=" + std::string(fs->mds_map.get_fs_name());
+	  mon_cap_string += " fsname=" + std::string(fs->mds_map.get_fs_name());
```

Comment:
This introduces a bug. This line of code should be moved outside the __for__ loop. Move it to line 1658.

Otherwise I hit the following issue,
$ ./bin/ceph fs authorize a client.test_a / r /volumes rw
Error EINVAL: mon capability parse failed, stopped at 'fsname=a' of 'allow r fsname=a fsname=a'

Reply:
@rishabh-d-dave please also add a test case for this ^.

Reply:
I've moved that line and fixed the bug. Thanks @ajarr for testing and pointing out to it!

Reply:
In regards to tests for `fs authorize` subcommand, I think I've missed out the fact that these tests are not truely multifs tests. Although caps produced by `fs authorize` are definitely affected by addition of "fsname" clause, but these caps are not multifs caps yet, so no multifs testing is really possible and, therefore, we should move `TestSubCmdFSAuthorize` to `test_admin.py`.

This also made me curious and made me note the following behaviour of `fs authorize` subcommand -
```
$ ./bin/ceph fs ls
name: a, metadata pool: cephfs.a.meta, data pools: [cephfs.a.data ]
name: cephfs2, metadata pool: cephfs2_meta, data pools: [cephfs2_data ]
$ ./bin/ceph fs authorize a client.someuser / rw
[client.someuser]
	key = AQAjeUNfcvezHhAAzFTLqmpzZRqgEV5bRReChw==
$ ./bin/ceph fs authorize cephfs2 client.someuser / rw
Error EINVAL: client.someuser already has fs capabilities that differ from those supplied. To generate a new auth key for client.someuser, first remove client.someuser from configuration files, execute 'ceph auth rm client.someuser', then execute this command again.
$
```

I think `fs authorize`'s behaviour should be modified so that running it a second time for second FS would assign the client the MDS caps for the second FS too. Once this is done, I should add `TestSubCmdFSAuthorize` to `test_multifs.py` too to test this feature.

@ajarr @batrick Does this sound good? If so, I am thinking to create a separate PR for this.

Reply:
Move TestSubCmdFsAuthorize to test_admin.

Reply:
> In regards to tests for `fs authorize` subcommand, I think I've missed out the fact that these tests are not truely multifs tests. Although caps produced by `fs authorize` are definitely affected by addition of "fsname" clause, but these caps are not multifs caps yet, so no multifs testing is really possible and, therefore, we should move `TestSubCmdFSAuthorize` to `test_admin.py`.
> 
> This also made me curious and made me note the following behaviour of `fs authorize` subcommand -
> 
> ```
> $ ./bin/ceph fs ls
> name: a, metadata pool: cephfs.a.meta, data pools: [cephfs.a.data ]
> name: cephfs2, metadata pool: cephfs2_meta, data pools: [cephfs2_data ]
> $ ./bin/ceph fs authorize a client.someuser / rw
> [client.someuser]
> 	key = AQAjeUNfcvezHhAAzFTLqmpzZRqgEV5bRReChw==
> $ ./bin/ceph fs authorize cephfs2 client.someuser / rw
> Error EINVAL: client.someuser already has fs capabilities that differ from those supplied. To generate a new auth key for client.someuser, first remove client.someuser from configuration files, execute 'ceph auth rm client.someuser', then execute this command again.
> $
> ```
> 
> I think `fs authorize`'s behaviour should be modified so that running it a second time for second FS would assign the client the MDS caps for the second FS too.

Great catch! One challenge with `fs authorize` is that it tries to be idempotent. If you run it again with a different set of caps then it complains. Also, there isn't really a clean way to "deauthorize" so chaining multiple calls for each fs is incompatible with the current interface. You'll want to think about this before making any changes. A tracker ticket is a good place to start the discussion.

> Once this is done, I should add `TestSubCmdFSAuthorize` to `test_multifs.py` too to test this feature.
> 
> @ajarr @batrick Does this sound good? If so, I am thinking to create a separate PR for this.

Good. Thanks.

Reply:
> > In regards to tests for `fs authorize` subcommand, I think I've missed out the fact that these tests are not truely multifs tests. Although caps produced by `fs authorize` are definitely affected by addition of "fsname" clause, but these caps are not multifs caps yet, so no multifs testing is really possible and, therefore, we should move `TestSubCmdFSAuthorize` to `test_admin.py`.
> > This also made me curious and made me note the following behaviour of `fs authorize` subcommand -
> > ```
> > $ ./bin/ceph fs ls
> > name: a, metadata pool: cephfs.a.meta, data pools: [cephfs.a.data ]
> > name: cephfs2, metadata pool: cephfs2_meta, data pools: [cephfs2_data ]
> > $ ./bin/ceph fs authorize a client.someuser / rw
> > [client.someuser]
> > 	key = AQAjeUNfcvezHhAAzFTLqmpzZRqgEV5bRReChw==
> > $ ./bin/ceph fs authorize cephfs2 client.someuser / rw
> > Error EINVAL: client.someuser already has fs capabilities that differ from those supplied. To generate a new auth key for client.someuser, first remove client.someuser from configuration files, execute 'ceph auth rm client.someuser', then execute this command again.
> > $
> > ```
> > 
> > 
> > I think `fs authorize`'s behaviour should be modified so that running it a second time for second FS would assign the client the MDS caps for the second FS too.
> 
> Great catch! One challenge with `fs authorize` is that it tries to be idempotent. If you run it again with a different set of caps then it complains. Also, there isn't really a clean way to "deauthorize" so chaining multiple calls for each fs is incompatible with the current interface. You'll want to think about this before making any changes. A tracker ticket is a good place to start the discussion.

https://tracker.ceph.com/issues/47264
https://tracker.ceph.com/issues/47266